### PR TITLE
SM/MM: added support for pending operations.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1023/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/449/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe can you please review?

**Comment by shamisp** [comment]
Autobot build

File: src/uct/sm/mm/mm_ep.c [hunk 1]
```diff
@@ -271,6 +275,68 @@ ssize_t uct_mm_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id, uct_pack_callback_t pack_
...
+{
+    uct_pending_req_t *req = ucs_container_of(elem, uct_pending_req_t, priv);
+    unsigned *send_attempts = arg;
+    ucs_status_t status;
+
+    status = req->func(req);
+    ucs_trace_data("progress pending request %p returned %s", req,
+                   ucs_status_string(status));
+
+    if (status == UCS_OK) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: pls add assert than *send_attempts > 0

File: src/uct/sm/mm/mm_iface.c [hunk 2]
```diff
@@ -248,8 +253,13 @@ static inline void uct_mm_iface_poll_fifo(uct_mm_iface_t *iface)
 static void uct_mm_iface_progress(void *arg)
 {
     uct_mm_iface_t *iface = arg;
+    unsigned send_attempts = iface->config.send_attempts;
 
+    /* progress receive */
     uct_mm_iface_poll_fifo(iface);
+
+    /* progress the pending sends (if there are any) */
+    ucs_arbiter_dispatch(&iface->arbiter, 1, uct_mm_ep_process_pending, &send_attempts);
```
Comment (shamisp) [inline src/uct/sm/mm/mm_iface.c]: I'm curious about the overhead of this check. Did you compare latency with and without the call ?

File: src/uct/sm/mm/mm_iface.c [hunk 3]
```diff
@@ -37,6 +37,10 @@ static ucs_config_field_t uct_mm_iface_config_table[] = {
      " try - Try to allocate memory using huge pages and if it fails, allocate regular pages.\n",
      ucs_offsetof(uct_mm_iface_config_t, hugetlb_mode), UCS_CONFIG_TYPE_TERNARY},
 
+    {"SEND_ATTEMPTS", "64",
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: this is not exactly "send attempts", this is how many endpoints to poll on in every progress iterations.
I think the default should be smaller (e.g 4), and also in runtime it should be min(config.poll_count, iface->num_eps)

File: src/uct/sm/mm/mm_iface.c (hunk 2 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: should be fast if arbiter is empty. this function is inline.

Comment (shamisp) [inline src/uct/sm/mm/mm_iface.c]: I would suggest to measure/validate this. Otherwise the PR looks good.

Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: we may also register this as an additional progress callback

Comment (shamisp) [inline src/uct/sm/mm/mm_iface.c]: since you check this on each iteration - yeah

Comment (shamisp) [inline src/uct/sm/mm/mm_iface.c]: for now I think I would prefer to keep it within mm progress

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/518/ for details (Mellanox internal link).

File: test/gtest/uct/test_pending.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    int i, counter = 0;
+
+    initialize();
+    check_caps(UCT_IFACE_FLAG_AM_SHORT | UCT_IFACE_FLAG_PENDING);
+
+    /* set a callback for the uct to invoke for receiving the data */
+    uct_iface_set_am_handler(m_e2->iface(), 0, pending_am_handler , &counter);
+
+    /* send the data until the resources run out */
+    for (i = 0; i < 100000; i++) {
```
Comment (yosefe) [inline test/gtest/uct/test_pending.cc]: pls do 100000 / ucs::test_time_multiplier() to make less iters with valgrind

File: test/gtest/uct/test_pending.cc [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+                req = (pending_send_request_t *) malloc(sizeof(*req));
+                req->buffer   = send_data;
+                req->hdr      = test_pending_hdr;
+                req->length   = sizeof(send_data);
+                req->ep       = m_e1->ep(0);
+                req->uct.func = pending_send_op;
+
+                status = uct_ep_pending_add(m_e1->ep(0), &req->uct);
+                if (status != UCS_OK) {
```
Comment (yosefe) [inline test/gtest/uct/test_pending.cc]: this doesn't meant it was sent, if add returns busy need to retry sending.

File: test/gtest/uct/test_pending.cc [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+        status = uct_ep_am_short(req->ep, 0, req->hdr, &req->buffer, req->length);
+        if (status == UCS_OK) {
+            free(req);
+        }
+        return status;
+    }
+
+    void cleanup() {
+        uct_test::cleanup();
```
Comment (yosefe) [inline test/gtest/uct/test_pending.cc]: no need to do it, if cleanup is not defined parent class cleanup will be called automatically

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/519/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1090/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1091/ for details.

File: test/gtest/uct/test_pending.cc (hunk 4 above)
Comment (alinask) [inline test/gtest/uct/test_pending.cc]: will do. 
but the point of this test is to do many iterations to make some requests go to pending. lowering this value can prevent it. 
i suppose the number for valgrind is still high enough...

**Comment by shamisp** [comment]
why pending was removed ?

**Comment by alinask** [comment]
the test currently doesnt pass with ud.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1092/ for details.

**Comment by shamisp** [comment]
Do we have open bug for this ?

**Comment by alinask** [comment]
no yet. this test wasn't pushed yet :)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/520/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1093/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/521/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1094/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/522/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:

**Pull request merged**